server/tests/unit/dataset/test_cxg_dataset.py

```diff
@@ -87,3 +89,90 @@ def test_tdb_bug_lossy(self):
         self.assertEqual(len(summary["columns"]), 1)
         self.assertEqual(len(summary["columns"][0]), 2638)
         self.assertEqual(summary["columns"][0].sum(), np.float32(-35.90116))
+
+    def test_save_obs_annotations_roundtrip(self):
+        data = self.get_data("pbmc3k.cxg")
+        user_id = "unit-test-user-roundtrip"
+        n_obs, _ = data.get_shape()
+        df = pd.DataFrame({"user_label": ["label"] * n_obs})
+
+        data.save_obs_annotations(df, user_id=user_id)
+        saved = data.get_saved_obs_annotations(user_id=user_id)
+        self.assertIsNotNone(saved)
+        # Compare values rather than exact DataFrame structure since TileDB converts to categorical
+        self.assertEqual(saved.columns.tolist(), df.columns.tolist())
+        self.assertEqual(saved["user_label"].values.tolist(), df["user_label"].values.tolist())
+
+        # Ensure persisted annotations can be reloaded via a new adaptor instance
+        data.cleanup()
+        reloaded = self.get_data("pbmc3k.cxg")
+        persisted = reloaded.get_saved_obs_annotations(user_id=user_id)
+        self.assertIsNotNone(persisted)
+        # Compare values rather than exact DataFrame structure since TileDB converts to categorical
+        self.assertEqual(persisted.columns.tolist(), df.columns.tolist())
+        self.assertEqual(persisted["user_label"].values.tolist(), df["user_label"].values.tolist())
+        reloaded.cleanup()
+
+    def test_save_gene_sets_tid_flow(self):
+        data = self.get_data("pbmc3k.cxg")
+        genesets = [
+            {
+                "geneset_name": "my_set",
+                "geneset_description": "",
+                "genes": [{"gene_symbol": "FOO", "gene_description": ""}],
+            }
+        ]
+
+        user_id = "unit-test-user-tid-flow"
+
+        data.save_gene_sets(genesets, tid=1, user_id=user_id)
```

Comment:
Nice coverage here: we assert TID monotonicity, persistence across reopen, and that the schema now reflects writable user columns. Handy sanity check for anyone touching the TileDB adapter.